### DISCUSSION TITLE
How to convert empty lines into <br> tag?

### DISCUSSION THREAD

**Discussion by marktopper** [thephpleague/commonmark#654] [Q&A] [answered]
How can I convert empty lines into a `<br>` tag?

Example:
```markdown
foo

bar
```
Turns into this:
```html
<p>Foo</p>
<p>Bar</p>
```
While I want this:
```html
<p>Foo</p>
<br />
<p>Bar</p>
```

**Comment by colinodell** [comment, marked as answer]
This behavior follows the CommonMark spec.  However, there are ways you can customize this library which might work.

If you only want this tag to appear after paragraphs, one possible option for you would be extending the `ParagraphRenderer` to always add a `<br>` after every paragraph:

```php
class MyParagraphRenderer extends ParagraphRenderer
{
    public function render(AbstractBlock $block, ElementRendererInterface $htmlRenderer, $inTightList = false)
    {
        return parent::render($block, $htmlRenderer, $inTightList) . '<br>';
    }
}
```

If you only want `<br>` to appear between paragraphs, you'll probably need to implement a `DocumentProcessor` which iterates through the tree of parsed elements and inserts a `<br>` when it finds a paragraph immediately followed by another.

If you truly want to convert ALL empty lines to `<br>`s you'll need to implement a block parser instead.  This _might_ cause unexpected side effects (like empty lines in code blocks converting to `<br>`s) - I'm not 100% sure though, you'd have to test it out.

Our docs should give you a good starting point on how to implement these various approaches: https://commonmark.thephpleague.com/customization/overview/

Hope that helps!

**Comment by froschdesign** [comment]
@marktopper 
The question is: why do you want the extra `br` element. If the `br` is needed for extra space then the element is wrong here. Use CSS instead.
A `br` element is used for a line break, not for a visual styling.

See also at MDN:

> Do not use `<br>` to increase the gap between lines of text; use the CSS margin property or the <p> element.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/br#Notes

**Comment by marktopper** [comment]
Thanks @colinodell, will try that out.

@froschdesign, the reason I need this is because that I have a Markdown editor in Javascript on my application's admin panel that makes an empty newline into a break tag.
Thansk for your feedback, I will try to see if I can change the behavoir on the Javascript markdown editor first and use this as last option.
